lambda/matching.ml

```diff
@@ -142,19 +142,82 @@ let all_record_args lbls =
 type 'a clause = 'a * lambda
 
 module Non_empty_clause = struct
-  type 'a t = ('a * pattern list) clause
+  type 'a t = ('a * Typedtree.pattern list) clause
 
   let of_initial = function
     | [], _ -> assert false
     | pat :: patl, act -> ((pat, patl), act)
+
+  let map_head f ((p, patl), act) = ((f p, patl), act)
 end
 
-module General = struct
-  type nonrec pattern = pattern
+type simple_view =
+  [ `Any
+  | `Constant of constant
+  | `Tuple of pattern list
+  | `Construct of Longident.t loc * constructor_description * pattern list
+  | `Variant of label * pattern option * row_desc ref
+  | `Record of
+    (Longident.t loc * label_description * pattern) list * closed_flag
+  | `Array of pattern list
+  | `Lazy of pattern ]
+
+type half_simple_view =
+  [ simple_view | `Or of pattern * pattern * row_desc option ]
+
+type general_view =
+  [ half_simple_view
+  | `Var of Ident.t * string loc
+  | `Alias of pattern * Ident.t * string loc ]
+
+module General : sig
+  type pattern = general_view pattern_data
 
   type clause = pattern Non_empty_clause.t
+
+  val view : Typedtree.pattern -> pattern
+
+  val erase : [< general_view ] pattern_data -> Typedtree.pattern
+end = struct
+  type pattern = general_view pattern_data
+
+  type clause = pattern Non_empty_clause.t
+
+  let view_desc = function
+    | Tpat_any -> `Any
+    | Tpat_var (id, str) -> `Var (id, str)
+    | Tpat_alias (p, id, str) -> `Alias (p, id, str)
+    | Tpat_constant cst -> `Constant cst
+    | Tpat_tuple ps -> `Tuple ps
+    | Tpat_construct (cstr, cstr_descr, args) ->
+        `Construct (cstr, cstr_descr, args)
+    | Tpat_variant (cstr, arg, row_desc) -> `Variant (cstr, arg, row_desc)
+    | Tpat_record (fields, closed) -> `Record (fields, closed)
+    | Tpat_array ps -> `Array ps
+    | Tpat_or (p, q, row_desc) -> `Or (p, q, row_desc)
+    | Tpat_lazy p -> `Lazy p
+
+  let view p : pattern = { p with pat_desc = view_desc p.pat_desc }
+
+  let erase_desc = function
+    | `Any -> Tpat_any
+    | `Var (id, str) -> Tpat_var (id, str)
+    | `Alias (p, id, str) -> Tpat_alias (p, id, str)
+    | `Constant cst -> Tpat_constant cst
+    | `Tuple ps -> Tpat_tuple ps
+    | `Construct (cstr, cst_descr, args) ->
+        Tpat_construct (cstr, cst_descr, args)
+    | `Variant (cstr, arg, row_desc) -> Tpat_variant (cstr, arg, row_desc)
+    | `Record (fields, closed) -> Tpat_record (fields, closed)
+    | `Array ps -> Tpat_array ps
+    | `Or (p, q, row_desc) -> Tpat_or (p, q, row_desc)
+    | `Lazy p -> Tpat_lazy p
+
+  let erase p = { p with pat_desc = erase_desc p.pat_desc }
 end
 
+let omega_ : [> `Any ] pattern_data = { Parmatch.omega with pat_desc = `Any }
```

Comment:
It may be nice to have a better qualifier than `_` for this `omega`. `General.omega`? `omega_view`? `just_omega`? `exactly_omega`?

Reply:
The names you give are not clearly better (`General.omega` is probably the best candidate). This declaration is gone in the final PR state of #9321, replaced by `Simple.omega : [> Simple.view ] pattern`, so I would propose to do nothing for now.

Reply:
Indeed, `omega_` is a fine ephemeral name.